// @webpack-cli/generators/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const entry_1 = require("./entry");
exports.entryQuestions = entry_1.default;
const languageSupport_1 = require("./languageSupport");
exports.langQuestionHandler = languageSupport_1.default;
exports.LangType = languageSupport_1.LangType;
exports.getBabelLoader = languageSupport_1.getBabelLoader;
exports.getTypescriptLoader = languageSupport_1.getTypescriptLoader;
const plugins_1 = require("./plugins");
exports.plugins = plugins_1.default;
exports.replaceAt = plugins_1.replaceAt;
exports.generatePluginName = plugins_1.generatePluginName;
const styleSupport_1 = require("./styleSupport");
exports.styleQuestionHandler = styleSupport_1.default;
exports.StylingType = styleSupport_1.StylingType;
exports.LoaderName = styleSupport_1.LoaderName;
exports.StyleRegex = styleSupport_1.StyleRegex;
const tooltip_1 = require("./tooltip");
exports.tooltip = tooltip_1.default;
const validate_1 = require("./validate");
exports.validate = validate_1.default;
const webpackConfig_1 = require("./webpackConfig");
exports.getDefaultOptimization = webpackConfig_1.getDefaultOptimization;
