// @polkadot/ui-settings/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Settings", {
  enumerable: true,
  get: function () {
    return _Settings.Settings;
  }
});
Object.defineProperty(exports, "settings", {
  enumerable: true,
  get: function () {
    return _Settings.settings;
  }
});
Object.defineProperty(exports, "ENDPOINT_DEFAULT", {
  enumerable: true,
  get: function () {
    return _index.ENDPOINT_DEFAULT;
  }
});
Object.defineProperty(exports, "ICON_DEFAULT", {
  enumerable: true,
  get: function () {
    return _index.ICON_DEFAULT;
  }
});
Object.defineProperty(exports, "ICON_DEFAULT_HOST", {
  enumerable: true,
  get: function () {
    return _index.ICON_DEFAULT_HOST;
  }
});
Object.defineProperty(exports, "LANGUAGE_DEFAULT", {
  enumerable: true,
  get: function () {
    return _index.LANGUAGE_DEFAULT;
  }
});
Object.defineProperty(exports, "LOCKING_DEFAULT", {
  enumerable: true,
  get: function () {
    return _index.LOCKING_DEFAULT;
  }
});
Object.defineProperty(exports, "PREFIX_DEFAULT", {
  enumerable: true,
  get: function () {
    return _index.PREFIX_DEFAULT;
  }
});
Object.defineProperty(exports, "UIMODE_DEFAULT", {
  enumerable: true,
  get: function () {
    return _index.UIMODE_DEFAULT;
  }
});
Object.defineProperty(exports, "UITHEME_DEFAULT", {
  enumerable: true,
  get: function () {
    return _index.UITHEME_DEFAULT;
  }
});
Object.defineProperty(exports, "packageInfo", {
  enumerable: true,
  get: function () {
    return _packageInfo.packageInfo;
  }
});
exports.default = void 0;

var _Settings = require("./Settings.cjs");

var _index = require("./defaults/index.cjs");

var _packageInfo = require("./packageInfo.cjs");

// Copyright 2017-2021 @polkadot/ui-settings authors & contributors
// SPDX-License-Identifier: Apache-2.0
var _default = _Settings.settings;
exports.default = _default;
