// @webpack-cli/generators/plugin-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
const path = require("path");
const addon_generator_1 = require("./addon-generator");
/**
 * A yeoman generator class for creating a webpack
 * plugin project. It adds some starter plugin code
 * and runs `webpack-defaults`.
 *
 * @class PluginGenerator
 * @extends {Generator}
 */
const PluginGenerator = addon_generator_1.default([
    {
        default: "my-webpack-plugin",
        filter: _.kebabCase,
        message: "Plugin name",
        name: "name",
        type: "input",
        validate: (str) => str.length > 0
    }
], path.resolve(__dirname, "..", "generate-plugin"), [
    "src/cjs.js.tpl",
    "test/test-utils.js.tpl",
    "test/functional.test.js.tpl",
    "examples/simple/src/index.js.tpl",
    "examples/simple/src/lazy-module.js.tpl",
    "examples/simple/src/static-esm-module.js.tpl"
], ["src/_index.js.tpl", "examples/simple/_webpack.config.js.tpl"], (gen) => ({ name: _.upperFirst(_.camelCase(gen.props.name)) }));
exports.default = PluginGenerator;
